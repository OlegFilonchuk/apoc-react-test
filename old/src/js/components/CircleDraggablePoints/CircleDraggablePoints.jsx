import React, { Component } from 'react';
import PropTypes from 'prop-types';

import DraggablePoints from '../DraggablePoints/DraggablePoints';

const halfPi = Math.PI / 2;
const doublePi = Math.PI * 2;

const round = number => Math.floor(number * 1000) / 1000;

const getPointPercentage = (oldPoints, updatedPoints, draggedPointId, center) => {
  const { x, y } = updatedPoints[draggedPointId];
  const dx = x - center.x;
  const dy = y - center.y;

  const shiftX = dx < 0 ? Math.PI : 0;
  const angle = shiftX + halfPi - Math.atan(dy / dx);

  const percentage = round(angle / doublePi);

  return percentage;
};

class CircleDraggablePoints extends Component {
  constructor(props) {
    super(props);

    this.pointsToPositions = this.pointsToPositions.bind(this);
    this.percentageToPosition = this.percentageToPosition.bind(this);
    this.onDraggablePointsUpdate = this.onDraggablePointsUpdate.bind(this);
  }

  onDraggablePointsUpdate(newPoints, draggedPointId) {
    const recalculatedPercentage = getPointPercentage(this.props.points, newPoints, draggedPointId, this.props.center);
    const points = this.props.points.slice();

    points[draggedPointId].percentage = recalculatedPercentage;

    this.props.update(points);
  }

  pointsToPositions(points) {
    return points.map(point => {
      const { name, percentage } = point;
      const { x, y } = this.percentageToPosition(percentage);

      return { name, x, y };
    });
  }

  percentageToPosition(percentage) {
    const { center, radius } = this.props;

    const angle = halfPi - doublePi * percentage;

    const x = center.x + radius * Math.cos(angle);
    const y = center.y + radius * Math.sin(angle);

    return { x, y };
  }

  render() {
    const points = this.pointsToPositions(this.props.points);

    return (
      <DraggablePoints
        points={points}
        update={this.onDraggablePointsUpdate}
        width={this.props.width}
        height={this.props.height}
        className={this.props.className}
      />
    );
  }
}

CircleDraggablePoints.propTypes = {
  /**
   * Center coordinates of circle
   * coordinates are calculated from left-bottom corner of container
   */
  center: PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number
  }).isRequired,

  /**
   * Radius of the circle
   */
  radius: PropTypes.number.isRequired,

  /**
   * Points percentages and names to display
   */
  points: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      percentage: PropTypes.number
    })
  ).isRequired,

  /**
   * Callback when points coordinates change. Passes new points array as an argument.
   */
  update: PropTypes.func.isRequired,

  /**
   * Width of container element
   */
  width: PropTypes.number.isRequired,

  /**
   * Height of container element
   */
  height: PropTypes.number.isRequired,

  /**
   * Any custom class names
   */
  className: PropTypes.string
};

CircleDraggablePoints.defaultProps = {
  className: ''
};

export default CircleDraggablePoints;
