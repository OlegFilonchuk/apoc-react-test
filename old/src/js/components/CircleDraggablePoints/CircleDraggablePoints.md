# Example
        
      const radius = 100;
      const initialState = {
        points: [
          {
            name: '',
            percentage: 0
          },
          {
            name: '',
            percentage: 0.25
          },
          {
            name: '',
            percentage: 0.7
          }
        ]
      }
      
      const center = {
        x: 250,
        y: 100
      }

      const update = (points) => {
        setState({points});
      };
      
      const containerStyle = {
        position: 'relative',
        backgroundColor: '#eee',
        width: 500,
        height: 200
      };

      const circleSyle = {
        position: 'absolute',
        top: containerStyle.height - center.y - radius,
        left: center.x - radius,
        width: radius * 2,
        height: radius * 2,
        WebkitBorderRadius: radius,
        MozBorderRadius: radius,
        border: '1px solid black'
      };

      (
        <div style={containerStyle}>
          <CircleDraggablePoints 
            points={state.points} 
            radius={radius} 
            center={center} 
            update={update}
            width={containerStyle.width}
            height={containerStyle.height}
          />
          <div style={circleSyle}></div>
        </div>
      )
