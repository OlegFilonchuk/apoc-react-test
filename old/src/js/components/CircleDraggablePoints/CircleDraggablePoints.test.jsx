import cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import { mount } from 'enzyme';

import CircleDraggablePoints from './CircleDraggablePoints';

const points = [
  {
    name: 'name1',
    percentage: 0
  },
  {
    name: 'name2',
    percentage: 0.5
  }
];

const center = {
  x: 50,
  y: 50
};
const radius = 50;
const width = 100;
const height = 100;

describe('<CircleDraggablePoints />', () => {
  const mountDraggablePoints = onUpdateSpy =>
    mount(
      <CircleDraggablePoints
        points={points}
        update={onUpdateSpy}
        center={center}
        radius={radius}
        width={width}
        height={height}
      />
    );

  it('should have proper structure', () => {
    const component = mountDraggablePoints(jest.fn());

    expect(component.html()).toMatchSnapshot();
  });

  describe('mouse movement', () => {
    const clickElement = element =>
      element.simulate('mousedown', {
        clientX: 0,
        clientY: 0
      });

    const moveMouse = (x, y) => {
      const body = window.document.body;
      const mouseMove = new CustomEvent('mousemove');

      mouseMove.clientX = x;
      mouseMove.clientY = y;

      body.dispatchEvent(mouseMove);
    };

    const updateSpy = jest.fn();

    const pointsComponents = mountDraggablePoints(updateSpy).find('.point__handler');

    const constructExpectedPoints = pointsPercentages => {
      const expectedPoints = cloneDeep(points);

      pointsPercentages.forEach((percentage, index) => {
        expectedPoints[index].percentage = percentage;
      });

      return expectedPoints;
    };

    const interactWithPoint = (point, mouseMoveX, mouseMoveY) => {
      clickElement(point);
      moveMouse(mouseMoveX, mouseMoveY);
    };

    afterEach(() => updateSpy.mockClear());

    it('should update first point percentage on mouse move', () => {
      const expectedPoints = constructExpectedPoints([0.148, 0.5]);

      const point = pointsComponents.first();

      interactWithPoint(point, 35, 24);

      expect(updateSpy).toHaveBeenCalledWith(expectedPoints);
    });

    it('should update second point percentage on mouse move', () => {
      const expectedPoints = constructExpectedPoints([0.148, 0.383]);

      const point = pointsComponents.last();

      interactWithPoint(point, 45, 14);

      expect(updateSpy).toHaveBeenCalledWith(expectedPoints);
    });
  });
});
