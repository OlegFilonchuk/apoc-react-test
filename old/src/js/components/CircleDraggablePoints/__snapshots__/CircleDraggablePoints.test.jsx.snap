// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircleDraggablePoints /> should have proper structure 1`] = `"<div class=\\"draggable-points\\"><div class=\\"point\\" style=\\"left: 50px; bottom: 100px;\\"><div class=\\"point__handler\\"></div></div><div class=\\"point\\" style=\\"left: 50px; bottom: 0px;\\"><div class=\\"point__handler\\"></div></div></div>"`;
